homeassistant/components/proximity/coordinator.py

```diff
@@ -0,0 +1,268 @@
+
+from dataclasses import dataclass
+import logging
+from typing import TypedDict
+
+from homeassistant.const import (
+    ATTR_LATITUDE,
+    ATTR_LONGITUDE,
+    CONF_DEVICES,
+    CONF_UNIT_OF_MEASUREMENT,
+    CONF_ZONE,
+    UnitOfLength,
+)
+from homeassistant.core import HomeAssistant, State
+from homeassistant.helpers.event import async_track_state_change
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+from homeassistant.util.location import distance
+from homeassistant.util.unit_conversion import DistanceConverter
+
+from .const import (
+    CONF_IGNORED_ZONES,
+    CONF_TOLERANCE,
+    DEFAULT_DIR_OF_TRAVEL,
+    DEFAULT_DIST_TO_ZONE,
+    DEFAULT_NEAREST,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class StateChangedData:
+    """StateChangedData class."""
+
+    entity_id: str
+    old_state: State | None
+    new_state: State | None
+
+
+class ProximityData(TypedDict):
+    """ProximityData type class."""
+
+    dist_to_zone: str | float
+    dir_of_travel: str | float
+    nearest: str | float
+
+
+class ProximityDataUpdateCoordinator(DataUpdateCoordinator[ProximityData]):
+    """Proximity data update coordinator."""
+
+    def __init__(
+        self, hass: HomeAssistant, friendly_name: str, config: ConfigType
+    ) -> None:
+        """Initialize the Proximity coordinator."""
+        self.ignored_zones: list[str] = config[CONF_IGNORED_ZONES]
+        self.proximity_devices: list[str] = config[CONF_DEVICES]
+        self.tolerance: int = config[CONF_TOLERANCE]
+        self.proximity_zone: str = config[CONF_ZONE]
+        self.unit_of_measurement: str = config.get(
+            CONF_UNIT_OF_MEASUREMENT, hass.config.units.length_unit
+        )
+        self.friendly_name = friendly_name
+
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=friendly_name,
+            update_interval=None,
+        )
+
+        self.data = {
+            "dist_to_zone": DEFAULT_DIST_TO_ZONE,
+            "dir_of_travel": DEFAULT_DIR_OF_TRAVEL,
+            "nearest": DEFAULT_NEAREST,
+        }
+
+        self.state_change_data: StateChangedData | None = None
+        async_track_state_change(
+            hass, self.proximity_devices, self.async_check_proximity_state_change
+        )
+
+    async def async_check_proximity_state_change(
+        self, entity: str, old_state: State | None, new_state: State | None
+    ) -> None:
+        """Fetch and process state change event."""
+        self.state_change_data = StateChangedData(entity, old_state, new_state)
+        await self.async_refresh()
+
+    async def _async_update_data(self) -> ProximityData:
+        """Calculate Proximity data."""
+        if self.state_change_data is None:
+            return self.data
+
+        if self.state_change_data.new_state is None:
+            _LOGGER.debug("no new_state -> abort")
+            return self.data
```

Comment:
i think this check could already be handled by the `async_check_proximity_state_change` above, to avoid unnecessary state change triggers to registered coordinator entities in case data are not changed anyway (_same counts below_)

Reply:
--> d88a9c9